_posts/2024-08-19-quarkus-metaprogramming.adoc

```diff
@@ -0,0 +1,372 @@
+
+.Flamegraph of out of the box Jackson’s serialization
+image::f1.png[]
+
+Further zooming the flamegraph on the Jackson’s `BeanPropertyWriter::serializeAsField` method and searching for the word “reflect” it is possible to see in how many different places, evidenced in purple in this flamegraph, Jackson needs to use Java reflection in order to read the state of the object to be serialized and write it into a json string.
+
+.Use of reflection in out of the box Jackson’s serialization
+image::f2.png[]
+
+As expected this use case requires a quite heavy use of reflection which is the thing that we want to avoid. We have already found that the json serialization is performed by the rest-jackson extension, so let’s see how we can modify it to achieve this goal.
+
+== Overriding Jackson serialization
+
+Jackson makes it possible to override its standard reflection-based serialization behavior in a pretty simple way. It is sufficient to implement a class extending its StdSerializer thus defining how an instance of a given pojo should be rendered in json, something like:
+
+[source, java]
+----
+public class PersonSerializer extends StdSerializer {
+    public PersonSerializer() {
+        super(Person.class);
+    }
+
+    public void serialize(Object obj, JsonGenerator jsonGen, SerializerProvider serProv) throws IOException {
+        Person person = (Person)obj;
+        jsonGen.writeStartObject();
+        jsonGen.writeNumberField("age", person.getAge());
+        jsonGen.writeStringField("name", person.getName());
+        jsonGen.writeEndObject();
+   }
+}
+----
+
+and then to add this serializer to a module and register it on the ObjectMapper used by Jackson to perform the json serialization.
+
+[source, java]
+----
+SimpleModule module = new SimpleModule();
+module.addSerializer(Person.class, new PersonSerializer());
+ObjectMapper mapper = new ObjectMapper();
+mapper.registerModule(module);
+----
+
+With this setup everytime Jackson will have to convert an instance of the Person class into its json representation it will do this through this custom serializer instead of accessing the person’s fields via reflection.
+
+The strategy to implement reflection-free json serialization is then quite straightforward, but of course it would be great if the Quarkus extension could generate and register those serializers automatically and effortlessly. The first step to achieve this is discovering at deployment time for which classes we need to generate those serializers, that in our case are the classes returned by all the REST endpoints in our application.
```

Comment:
You should probably note, here, that the approach considered in this blog is focused on JSON serialization for REST endpoints but that the approach could be extended to a wider scope.

Reply:
Done.